theano/gof/wrapper.py

```diff
@@ -0,0 +1,439 @@
+            void extract_%(attribute_name)s(PyObject* py_%(attribute_name)s) {
+                %(extract_code)s
+            }
+            """ % {
+                'attribute_name': attribute_name,
+                'extract_code': type_instance.c_extract(attribute_name, sub)
+            })
+
+        struct_declare = '\n'.join(c_declare_list)
+        struct_init = '\n'.join(c_init_list)
+        struct_cleanup = '\n'.join(c_cleanup_list)
+        struct_extract = '\n\n'.join(c_extract_list)
+        struct_extract_method = """
+        void extract(PyObject* object, int field_pos) {
+            switch(field_pos) {
+                // Extraction cases.
+                %s
+                // Default case.
+                default:
+                    PyErr_Format(PyExc_TypeError, "Wrapper: no extraction defined for a field %%d.", field_pos);
+                    this->setErrorOccurred();
+                    break;
+            }
+        }
+        """ % ('\n'.join(
+            [('case %d: extract_%s(object); break;' % (i, self.fields[i])) for i in range(self.length)])
+        )
+        return """
+        #ifndef %(struct_name_defined)s
+        #define %(struct_name_defined)s
+        struct %(struct_name)s {
+            /* Attributes, */
+            int %(struct_name)s_error;
+            %(struct_declare)s
+
+            /* Constructor. */
+            %(struct_name)s() {
+                %(struct_name)s_error = 0;
+                %(struct_init)s
+            }
+
+            /* Destructor. */
+            ~%(struct_name)s() {
+                // cleanup() is defined below.
+                cleanup();
+            }
+
+            /* Cleanup method. */
+            void cleanup() {
+                %(struct_cleanup)s
+            }
+
+            /* Extraction methods. */
+            %(struct_extract)s
+
+            /* Extract method. */
+            %(struct_extract_method)s
+
+            /* Other methods. */
+            void setErrorOccurred() {
+                ++%(struct_name)s_error;
+            }
+            int errorOccurred() {
+                return %(struct_name)s_error;
+            }
+        };
+        #endif
+        """ % locals()
+
+    def c_code_cache_version(self):
+        wrapper_c_code_version = (1, 6)
+        return (wrapper_c_code_version, tuple(t.c_code_cache_version() for t in self.types))
+
+    # As this struct has constructor and destructor, it could be instanciated on stack,
+    # but current implementations of C ops will then pass the instance by value at functions,
+    # so it's better to work directly with pointers.
+
+    def c_declare(self, name, sub, check_input=True):
+        struct_name = self.name
+        return """
+        %(struct_name)s* %(name)s;
+        """ % locals()
+
+    def c_init(self, name, sub):
+        # NB: It seems c_init() is not called for an op param.
+        # So the real initialization is done at top of c_extract.
+        return """
```

Comment:
One solution to all this would be to declare the struct directly rather than as a pointer.  Then it is always default-initialized on declaration and we don't need c_init().